va/va.go

```diff
@@ -29,96 +39,177 @@ func userAgent() string {
 		userAgentBase, runtime.GOOS, runtime.GOARCH)
 }
 
+type vaTask struct {
+	Identifier   string
+	Challenge    *core.Challenge
+	Registration *core.Registration
+	Results      chan core.ValidationRecord
+}
+
 type VAImpl struct {
 	log      *log.Logger
 	clk      clock.Clock
 	httpPort int
+	tasks    chan *vaTask
+	ca       *ca.CAImpl
 }
 
-func New(log *log.Logger, clk clock.Clock, httpPort int) *VAImpl {
-	return &VAImpl{
+func New(log *log.Logger, clk clock.Clock, httpPort int, ca *ca.CAImpl) *VAImpl {
+	va := &VAImpl{
 		log:      log,
 		clk:      clk,
 		httpPort: httpPort,
+		tasks:    make(chan *vaTask, taskQueueSize),
+		ca:       ca,
 	}
+
+	go va.processTasks()
+	return va
 }
 
-func (va VAImpl) Validate(identifier string, chal *core.Challenge, reg *core.Registration) error {
-	// TODO(@cpu): Implement validation for DNS-01, TLS-SNI-02, etc
-	var prob *acme.ProblemDetails
-	switch chal.Type {
-	case acme.ChallengeHTTP01:
-		prob = va.validateHTTP01(identifier, chal, reg)
-	default:
-		return fmt.Errorf("Invalid challenge type: %q", chal.Type)
+func (va VAImpl) ValidateChallenge(ident string, chal *core.Challenge, reg *core.Registration) {
+	task := &vaTask{
+		Identifier:   ident,
+		Challenge:    chal,
+		Registration: reg,
+		Results:      make(chan core.ValidationRecord, concurrentValidations),
 	}
+	// Submit the task for validation
+	va.tasks <- task
+}
 
-	authz := chal.Authz
-	now := va.clk.Now().UTC()
-	// Update the validated date for the challenge regardless of good/bad outcome
-	chal.ValidatedDate = now
-	chal.Validated = chal.ValidatedDate.Format(time.RFC3339)
-	if prob != nil {
-		// Update the challenge Error
-		chal.Error = prob
-		// Set the challenge and authorization to invalid
-		chal.Status = acme.StatusInvalid
-		authz.Status = acme.StatusInvalid
-		va.log.Printf("authz %s set INVALID by completed challenge %s", authz.ID, chal.ID)
-	} else {
+func (va VAImpl) processTasks() {
+TaskLoop:
+	for {
+		// Pull a task off of the tasks channel
+		task, ok := <-va.tasks
+		if !ok {
+			break
+		}
+		va.log.Printf("Pulled a task from the Tasks queue: %#v", task)
+		va.log.Printf("Starting %d validations.", concurrentValidations)
+
+		var wg sync.WaitGroup
```

Comment:
This effectively serializes the tasks; If one authorization is validating, another authorization can't be validating at the same time. I think that might potentially mask some client bugs. Also, this loop seems a little long, and I think the goto may not be necessary.

What if instead, this was a very short loop:

```
for task := range va.tasks {
  go process(task)
}
```

And the bulk of this function could be moved into `process(..)`.